test/src/test/java/org/corfudb/runtime/view/AddressSpaceViewTest.java

```diff
@@ -169,30 +166,88 @@ public void ensureStripingReadAllWorks()
+        rt.getAddressSpaceView().write(token, testPayload);
 
-        assertThat(r.getAddressSpaceView().read(ADDRESS_0).getPayload(getRuntime()))
+        assertThat(rt.getAddressSpaceView().read(ADDRESS_0).getPayload(getRuntime()))
                 .isEqualTo("hello world".getBytes());
 
         Range range = Range.closed(ADDRESS_0, ADDRESS_2);
         ContiguousSet<Long> addresses = ContiguousSet.create(range, DiscreteDomain.longs());
 
-        Map<Long, ILogData> m = r.getAddressSpaceView().read(addresses);
+        Map<Long, ILogData> m = rt.getAddressSpaceView().read(addresses);
 
         assertThat(m.get(ADDRESS_0).getPayload(getRuntime()))
                 .isEqualTo("hello world".getBytes());
-        assertThat(m.get(ADDRESS_1).isHole());
-        assertThat(m.get(ADDRESS_2).isHole());
+        assertThat(m.get(ADDRESS_1).isHole()).isTrue();
+        assertThat(m.get(ADDRESS_2).isHole()).isTrue();
+    }
+
+    /**
+     * Test bulk read can query the correct log unit server in case
+     * the requested addresses are stripped and span segments.
+     */
+    @Test
+    public void testMultiReadSpansSegments() {
+        final long segmentEnd = 5L;
+        addServer(SERVERS.PORT_0);
+        addServer(SERVERS.PORT_1);
+        addServer(SERVERS.PORT_2);
+
+        bootstrapAllServers(new TestLayoutBuilder()
+                .setEpoch(1L)
+                .addLayoutServer(SERVERS.PORT_0)
+                .addSequencer(SERVERS.PORT_0)
+                .buildSegment()
+                .setStart(0L)
+                .setEnd(segmentEnd)
+                .buildStripe()
+                .addLogUnit(SERVERS.PORT_0)
+                .addToSegment()
+                .buildStripe()
+                .addLogUnit(SERVERS.PORT_1)
+                .addToSegment()
+                .addToLayout()
+                .buildSegment()
+                .setStart(segmentEnd)
+                .setEnd(-1L)
+                .buildStripe()
+                .addLogUnit(SERVERS.PORT_1)
+                .addToSegment()
+                .buildStripe()
+                .addLogUnit(SERVERS.PORT_2)
+                .addToSegment()
+                .addToLayout()
+                .build());
+
+        // Shutdown management server to prevent segment merge
+        getManagementServer(SERVERS.PORT_0).shutdown();
+        getManagementServer(SERVERS.PORT_1).shutdown();
+        getManagementServer(SERVERS.PORT_2).shutdown();
+
+        CorfuRuntime rt = getRuntime().connect();
+
+        final String testString = "hello world ";
+
+        final long numAddresses = 10L;
+        for (long i = 0L; i < numAddresses; i++) {
+            TokenResponse token = rt.getSequencerView().next();
+            rt.getAddressSpaceView().write(token, (testString + i).getBytes());
+        }
+
+        Map<Long, ILogData> readResult = rt.getAddressSpaceView().read(
+                ContiguousSet.create(Range.closed(0L, numAddresses - 1), DiscreteDomain.longs()));
```

Comment:
Disable caching on the client and also test fetchAll. Also, `waitForWrite` option doesn't seem to be directly tested.

Reply:
Done.